text/0000-inline-buildpack.md

```diff
@@ -0,0 +1,122 @@
+
+We propose two new keys in the `[[build.buildpacks]]` table in `project.toml`:
+
+- `inline` - the contents of a script that will be run as the `bin/build` of the inline buildpack
+- `shell` - defines the shebang line of the script in `inline`
+
+(Note: there is no need to implement `bin/detect`. The detect phase will always pass because there's no foreseeable reason that a user would create an inline buildpack that they don't want to use)
+
+# How it Works
+[how-it-works]: #how-it-works
+
+When an entry in the `[[build.buildpacks]]` table contains an `inline` value, the platform WILL NOT attempt to download the buildpack `id@version`. Instead, it will create an ephemeral buildpack, with the following:
+
+- a `bin/detect` that always passes (i.e. `exit 0`)
+- a `bin/build` that contains the value of `inline` script without any changes (i.e. no magic)
+- a `buildpack.toml` using the `id` and `version` from the buildpack entry in `project.toml`
+
+When the ephemeral buildpack is executed, its working directory will contain the artifacts in the app repo. In this way, an `inline` script MAY reference other scripts in the repo. For example, a `lib/utils.sh` may be sourced by an inline Bash script.
+
+## Example
+
+The following `project.toml` creates a buildpack that will run two Rake tasks after running JVM and Ruby buildpacks.
+
+```toml
+[project]
+id = "my-app"
+
+[[build.buildpacks]]
+id = "example/jvm"
+version = "1.0"
+
+[[build.buildpacks]]
+id = "example/ruby"
+version = "1.0"
+
+[[build.buildpacks]]
+id = "me/rake-tasks"
+inline = """
+  rake war
+  rake db:migrate
+"""
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- It discourages buildpack reuse (i.e. if it's really easy to create a one-off buildpack, people are less like to make reuseable buildpacks and share them)
+- It encourages the copy-paste problems created by `Dockerfile`
+
+# Alternatives
+[alternatives]: #alternatives
+
+- https://github.com/buildpacks/rfcs/pull/17
+- [Packfile](https://github.com/sclevine/packfile/)
+
+# Prior Art
+[prior-art]: #prior-art
+
+- `Dockerfile`
+- [Packfile](https://github.com/sclevine/packfile/)
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- Can an inline buildpack be published to a registry?
+- Can inline buildpacks be root buildpacks?
+- Can a `project.toml` contain more than one inline buildpack?
+- Should `version` be required or should it default to `0.0.0`?
+- How will platforms implement this behavior?
+    - A future `/lifecycle/prepare` could generate the temporary buildpack from the inline script.
+
+# Spec. Changes (OPTIONAL)
+[spec-changes]: #spec-changes
+
+The [Project Descriptor extension](https://github.com/buildpacks/spec/blob/master/extensions/project-descriptor.md) will be ammended to include:
+
+## [[build.buildpacks]]
+
+The build table MAY contain an array of buildpacks. The schema for this table is:
+
+```toml
+[[build.buildpacks]]
```

Comment:
does it make sense to use an inline table since it's a separate concept and make it distinct.

```toml
[[build.buildpacks]]
shell = "<string (optional default=/bin/sh)>"
inline = "<script contents (optional)>"
```

Reply:
Actually it would be `[build.buildpacks.inline]` (not an array), which is one of the weirder things in TOML. I'd rather avoid it for that reason, but I admit this kinda makes sense. In addition to `shell` and `script`, we could also add `api` here and then not worry about those keys polluting the top-level `[[build.buildpacks]]`.

That said, since most of these are will now be mutually exclusive (i.e. only one of `version`, `uri`, `inline` are allowed), I wonder if we even care about polluting the top-level `[[build.buildpacks]]`.

Reply:
For most inline buildpacks, it would probably look like:

```toml
[[build.buildpacks]]
[build.buildpacks.inline]
api = "0.3"
script = "rake task"
```

I find that pretty dang cumbersome

Reply:
I don't actually find that that cumbersome especially b/c it makes it clearer what is happening, but not a hill I'm going to die on.

Reply:
Didn't we talk about using `api` and `id`?

```toml
[[build.buildpacks]]
id = "inliner"
api = "0.3"
[build.buildpacks.inline]
shell = "/bin/bash"
script = "rake task"
```

Reply:
`api` shouldn't go in `[build.buildpacks]` if we have the `inline` table. And requiring `id` just feels like making the user configure something they're never going to need. The `inline` table also doesn't solve the conflict between `uri` and `version`, so it's a half measure. 

Ultimately, the "table within an array of tables" syntax is the ugliest part of TOML, and I think it's best to avoid it unless we really don't have a choice.

Reply:
we'd still have to figure out what to do with this (which is invalid):

```toml
[[build.buildpacks]]
id = "inliner"
api = "0.3"
version = "1.0"
uri = "docker://gcr.io/projectriff/node-function:0.6.1"

[build.buildpacks.inline]
shell = "/bin/bash"
script = "rake task"
```

I feel like the `inline` table doesn't really solve the problem, unless we force the other attributes into their own table.

Reply:
> The inline table also doesn't solve the conflict between `uri` and `version`, so it's a half measure.
well, that's something we could solve, but I think maybe out of scope of this RFC. Ideally, I'd like this to not compound the problem. `uri` and `version` are the common case, so I expect that's fine to have that at the top level. `inline` is a special case to some degree. If we make `api` a valid key, is that confusing to users? What would a user expect if I define `api` for a non inline buildpack since it applies to a thing in `buildpack.toml`?

> And requiring id just feels like making the user configure something they're never going to need.
I pulled `id` from your example above.

> I think it's best to avoid it unless we really don't have a choice.
I basically don't agree with this.

```toml
[[build.buildpacks]]
id = "inliner-buildpack"
[build.buildpacks.inline]
api = "0.3"
shell = "/bin/bash"
script = "rake task"
```

If you have a one liner script you could even do:

```toml
[[build.buildpacks]]
id = "inliner-buildpack"
inline = {api = "0.3", shell = "/bin/bash", script = "rake task"}
```

I think technically you can even multiline the script for inline, but that's discouraged in toml land.

Reply:
My strongest opinion is that `inline` should be the field name for the literal inline string that specifies the code. That way `script` or `file` could be introduced later to point a file in the app directory that contains the code.